crates/ethereum-rpc/src/subscription.rs

```diff
@@ -0,0 +1,140 @@
+use citrea_evm::{log_matches_filter, Evm, Filter, LogResponse};
+use jsonrpsee::{SubscriptionMessage, SubscriptionSink};
+use reth_rpc_types::{BlockNumberOrTag, RichBlock};
+use sov_modules_api::WorkingSet;
+use tokio::sync::broadcast;
+
+pub(crate) struct SubscriptionManager {
+    new_heads_tx: broadcast::Sender<RichBlock>,
+    logs_tx: broadcast::Sender<Vec<LogResponse>>,
+}
+
+impl SubscriptionManager {
+    pub(crate) fn new<C: sov_modules_api::Context>(
+        storage: C::Storage,
+        soft_confirmation_rx: broadcast::Receiver<u64>,
+    ) -> Self {
+        let new_heads_tx = broadcast::channel(16).0;
+        let logs_tx = broadcast::channel(16).0;
+        let manager = Self {
+            new_heads_tx: new_heads_tx.clone(),
+            logs_tx: logs_tx.clone(),
+        };
+
+        let mut soft_confirmation_rx = soft_confirmation_rx;
+        // Spawn the task that will listen for new soft confirmation heights
+        // and send the corresponding ethereum block to subscribers
+        tokio::spawn(async move {
+            let evm = Evm::<C>::default();
+            loop {
+                let Ok(height) = soft_confirmation_rx.recv().await else {
+                    return;
+                };
+
+                let mut working_set = None;
+
+                if new_heads_tx.receiver_count() != 0 {
```

Comment:
pr has been going on for too long so let's create a new issue for separate this senders for different things

Reply:
wdym? new_heads and logs subscriptions are already separated.